File: mobile/lib/features/financial/domain/model/financial_event.dart
class FinancialEvent {
  final String id;
  final String userId;
  final double amount;
  final String currency;
  final String merchantName;
  final String? normalizedMerchantName;
  final double? merchantConfidence;
  final String financialEventType;
  final String category;
  final DateTime? paidAt;
  final DateTime? eventDateLocal;
  final String timezone;
  final String status;
  final String sourceProvider;
  final String? providerEventId;
  final String? sourceAccountIdMasked;
  final String? rawReference;
  final String? description;
  final String? locationText;
  final String? countryCode;
  final bool isReadOnly;
  final String? consentId;

  const FinancialEvent({
    required this.id,
    required this.userId,
    required this.amount,
    required this.currency,
    required this.merchantName,
    required this.normalizedMerchantName,
    required this.merchantConfidence,
    required this.financialEventType,
    required this.category,
    required this.paidAt,
    required this.eventDateLocal,
    required this.timezone,
    required this.status,
    required this.sourceProvider,
    required this.providerEventId,
    required this.sourceAccountIdMasked,
    required this.rawReference,
    required this.description,
    required this.locationText,
    required this.countryCode,
    required this.isReadOnly,
    required this.consentId,
  });

  String get displayMerchantName {
    final value = merchantName.trim();
    if (value.isNotEmpty) return value;

    final fallback = normalizedMerchantName?.trim() ?? '';
    return fallback.isNotEmpty ? fallback : 'Unknown merchant';
  }

  bool get isOutgoingType {
    switch (financialEventType.toUpperCase()) {
      case 'PURCHASE':
      case 'CASH_OUT':
      case 'FEE':
        return true;
      default:
        return false;
    }
  }
}

File: backend/src/main/java/com/lifeos/backend/financial/domain/FinancialEvent.java
package com.lifeos.backend.financial.domain;

import com.lifeos.backend.common.base.BaseEntity;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;

import java.math.BigDecimal;
import java.time.Instant;
import java.time.LocalDate;
import java.util.UUID;

@Entity
@Table(
        name = "financial_events",
        indexes = {
                @Index(name = "idx_financial_events_user_date", columnList = "userId,eventDateLocal"),
                @Index(name = "idx_financial_events_user_paidAt", columnList = "userId,paidAt"),
                @Index(name = "idx_financial_events_provider_event", columnList = "userId,providerEventId", unique = true)
        }
)
@Getter
@Setter
public class FinancialEvent extends BaseEntity {

    @Column(nullable = false)
    private UUID userId;

    @Column(nullable = false, precision = 18, scale = 4)
    private BigDecimal amount;

    @Column(nullable = false, length = 3)
    private String currency;

    @Column(nullable = false)
    private String merchantName;

    private String normalizedMerchantName;

    private Double merchantConfidence;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private FinancialEventType financialEventType;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private FinancialCategory category;

    @Column(nullable = false)
    private Instant paidAt;

    @Column(nullable = false)
    private LocalDate eventDateLocal;

    @Column(nullable = false)
    private String timezone;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private FinancialEventStatus status;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private SourceProvider sourceProvider;

    private String providerEventId;
    private String sourceAccountIdMasked;
    private String rawReference;

    @Column(length = 1000)
    private String description;

    private String locationText;
    private String countryCode;

    @Column(nullable = false)
    private Boolean isReadOnly = true;

    private String consentId;
}